### PR TITLE
Made bitmap loggable and climb changes

### PR DESCRIPTION
**Pull request by SumedhP**

### CODE REVIEW

File: src/main/java/frc/team2412/robot/commands/intake/IntakeBitmapCommand.java [hunk 1]
```diff
@@ -19,24 +21,24 @@
...
-        C(true, false, false, false, INTAKE_IN_SPEED, INDEX_IN_SPEED, INDEX_IN_SPEED, true), // wrong ball in ingest
-        D(false, true, false, true, INTAKE_IN_SPEED, INDEX_IN_SPEED, 0, false), // correct ball in feeder
-        E(false, true, false, false, INTAKE_IN_SPEED, INDEX_IN_SPEED, INDEX_IN_SPEED, true), // wrong ball in feeder
-        F(true, true, true, true, 0, 0, 0, false), // correct ball in both
-        G(true, true, false, true, INTAKE_OUT_SPEED, INDEX_OUT_SPEED, 0, false), // wrong ball in ingest, corect ball in
-                                                                                    // shooter (may change later)
-        H(true, true, true, false, 0, INDEX_IN_SPEED, INDEX_IN_SPEED, true), // correct ball in ingest, wrong ball in
-                                                                                // feeder
-        I(true, true, false, false, 0, INDEX_IN_SPEED, INDEX_IN_SPEED, true); // wrong ball in both
+        A(false, false, false, false, INTAKE_IN_SPEED, INDEX_IN_SPEED, INDEX_IN_SPEED, false, "no balls in system"), //
```
Comment (TAKBS2412) [inline src/main/java/frc/team2412/robot/commands/intake/IntakeBitmapCommand.java:24]: Is there a purpose for the empty comments? Or can they be safely deleted?

File: src/main/java/frc/team2412/robot/commands/intake/IntakeBitmapCommand.java [hunk 2]
```diff
@@ -19,24 +21,24 @@
...
+        D(false, true, false, true, INTAKE_IN_SPEED, INDEX_IN_SPEED, 0, false, "Correct ball in feeder"), //
+        E(false, true, false, false, INTAKE_IN_SPEED, INDEX_IN_SPEED, INDEX_IN_SPEED, true, "Wrong ball in feeder"), //
+        F(true, true, true, true, 0, 0, 0, false, "Correct ball in both"), //
+        G(true, true, false, true, INTAKE_OUT_SPEED, INDEX_OUT_SPEED, 0, false, "Wrong ingest, correct feeder"), //
+        H(true, true, true, false, 0, INDEX_IN_SPEED, INDEX_IN_SPEED, true, "Correct ingest, wrong feeder"), //
+        I(true, true, false, false, 0, INDEX_IN_SPEED, INDEX_IN_SPEED, true, "Wrong in both");//
 
         boolean ingestSensor, feederSensor, ingestColor, feederColor, shooterMisfire;
         double intakeMotorSpeed, ingestMotorSpeed, feederMotorSpeed;
+        String state;
```
Comment (TAKBS2412) [inline src/main/java/frc/team2412/robot/commands/intake/IntakeBitmapCommand.java:36]: This (and the other fields in the `Bitmap` enum) should be private to avoid unwanted accesses.

File: src/main/java/frc/team2412/robot/commands/intake/IntakeBitmapCommand.java [hunk 3]
```diff
@@ -61,6 +68,8 @@ public boolean equals(boolean ingestSensor, boolean feederSensor, boolean ingest
     private ShooterSubsystem shooterSubsystem;
     private ShooterVisionSubsystem shooterVisionSubsystem;
 
+    String currentState;
```
Comment (TAKBS2412) [inline src/main/java/frc/team2412/robot/commands/intake/IntakeBitmapCommand.java:71]: Is this variable being used?

**Review by TAKBS2412 — CHANGES_REQUESTED** [review]

File: src/main/java/frc/team2412/robot/commands/intake/IntakeBitmapCommand.java [hunk 4]
```diff
@@ -69,39 +81,40 @@ public IntakeBitmapCommand(IntakeSubsystem intakeSubsystem, IndexSubsystem index
...
-                    ShooterDataDistancePoint shooterData = ShooterConstants.dataPoints.getInterpolated(distance);
-                    shooterSubsystem.setHoodAngle(shooterData.getAngle());
-                    shooterSubsystem.setFlywheelRPM(shooterData.getRPM());
-                }
+                currentState = value;
+                break;
             }
         }
+
+        indexSubsystem.setBitmapState(currentState);
```
Comment (bobtseattle) [inline src/main/java/frc/team2412/robot/commands/intake/IntakeBitmapCommand.java:102]: If we are worried about overloading the CAN bus, we can consider comparing current state with previous state and skipping the set commands if no change detected.  But, this might be better in a separate commit.

File: src/main/java/frc/team2412/robot/subsystem/ClimbSubsystem.java [hunk 5]
```diff
@@ -208,4 +239,27 @@ public void simulationPeriodic() {
...
         lastUpdatedTime = timeNow;
     }
+
+    @Log
+    public double encoderPOsition() {
+        return climbFixedMotor.getSelectedSensorPosition();
+    }
+
+    @Config
+    public void resetEncoder(boolean hfh) {
```
Comment (TAKBS2412) [inline src/main/java/frc/team2412/robot/subsystem/ClimbSubsystem.java:249]: Please rename `hfh` to something more descriptive.

File: src/main/java/frc/team2412/robot/subsystem/ClimbSubsystem.java [hunk 6]
```diff
@@ -208,4 +239,27 @@ public void simulationPeriodic() {
...
+    @Config(name = "PID")
+    private void setPID(@Config(name = "P", defaultValueNumeric = P) double p,
+            @Config(name = "I", defaultValueNumeric = I) double i,
+            @Config(name = "D", defaultValueNumeric = D) double d) {
+        climbFixedMotor.config_kP(0, p);
+        climbFixedMotor.config_kI(0, i);
+        climbFixedMotor.config_kD(0, d);
+    }
+
+    // spotless:on
```
Comment (TAKBS2412) [inline src/main/java/frc/team2412/robot/subsystem/ClimbSubsystem.java:264]: Spotless should already be on, this can be deleted (looks like a remnant from disabling spotless earlier).

File: src/main/java/frc/team2412/robot/subsystem/ClimbSubsystem.java [hunk 7]
```diff
@@ -208,4 +239,27 @@ public void simulationPeriodic() {
         climbDynamicMotor.getSimCollection().setIntegratedSensorRawPosition((int) (motorDynamicSpeed / timeElapsed));
         lastUpdatedTime = timeNow;
     }
+
+    @Log
+    public double encoderPOsition() {
```
Comment (TAKBS2412) [inline src/main/java/frc/team2412/robot/subsystem/ClimbSubsystem.java:244]: Nit: should be `encoderPosition()`, not `encoderPOsition()`

File: src/main/java/frc/team2412/robot/subsystem/ClimbSubsystem.java [hunk 8]
```diff
@@ -170,15 +201,15 @@ public void stopAngledArm() {
...
-        }
+        // if (enabled) {
+        // System.out.println("motor set: " + value);
+        // if (Robot.isSimulation()) {
+        // motor.getSimCollection().setIntegratedSensorVelocity((int) value);
+        // } else {
+        motor.set(ControlMode.Position, value);
+        // }
+        // motor.set(ControlMode.PercentOutput, 1); simulator only
+        // }
```
Comment (TAKBS2412) [inline src/main/java/frc/team2412/robot/subsystem/ClimbSubsystem.java:212]: The commented out code here can be removed.

File: src/main/java/frc/team2412/robot/subsystem/ClimbSubsystem.java [hunk 9]
```diff
@@ -146,11 +177,11 @@ public void angleClimbHook(DoubleSolenoid.Value value) {
     }
 
     public void extendFixedArm() {
-        setMotor(testSpeedExtend.getValue().getDouble(), climbFixedMotor);
+        setMotor(30 * ENCODER_TICKS_PER_INCH, climbFixedMotor);
     }
 
     public void retractFixedArm() {
-        setMotor(testSpeedRetract.getValue().getDouble(), climbFixedMotor);
+        setMotor(2 * ENCODER_TICKS_PER_INCH, climbFixedMotor);
```
Comment (TAKBS2412) [inline src/main/java/frc/team2412/robot/subsystem/ClimbSubsystem.java:184]: This and the `30` in `extendFixedArm()` should be constants and optionally configurable in Shuffleboard.

File: src/main/java/frc/team2412/robot/subsystem/ClimbSubsystem.java [hunk 10]
```diff
@@ -80,21 +99,25 @@ public ClimbSubsystem(WPI_TalonFX climbFixedMotor, WPI_TalonFX climbDynamicMotor
...
         motorConfig.reverseSoftLimitEnable = false;
-        motorConfig.forwardSoftLimitThreshold = MAX_ENCODER_TICKS;
+        motorConfig.forwardSoftLimitThreshold = ARM_REACH_DISTANCE;
         motorConfig.reverseSoftLimitThreshold = MIN_ENCODER_TICKS;
         motorConfig.supplyCurrLimit = MOTOR_CURRENT_LIMIT;
 
         climbFixedMotor.configAllSettings(motorConfig);
+        climbFixedMotor.setNeutralMode(NeutralMode.Brake);
+
+        climbFixedMotor.config_kP(0, P);
```
Comment (TAKBS2412) [inline src/main/java/frc/team2412/robot/subsystem/ClimbSubsystem.java:114]: Should create a constant for the PID slot for here as well as in `setPID()`.

**Review by TAKBS2412 — CHANGES_REQUESTED** [review]
Mostly nit-picky style comments, not as important (but still worth looking at).

**Review by TAKBS2412 — APPROVED** [review]
Will merge once build passes

File: src/main/java/frc/team2412/robot/subsystem/ClimbSubsystem.java [hunk 11]
```diff
@@ -46,21 +46,21 @@
...
-        public static final double MID_RUNG_HEIGHT_INCH = 30;
-        public static final double RETRACT_HEIGHT_INCH = 2;
+        public static final double MID_RUNG_HEIGHT_INCH = 31;
+        public static final double RETRACT_HEIGHT_INCH = 20;
 
         public static final double CLIMB_OFFSET_INCHES = 28.5;
 
         public static final SupplyCurrentLimitConfiguration MOTOR_CURRENT_LIMIT = new SupplyCurrentLimitConfiguration(
-                true, 40, 40, 500);
+                true, 60, 60, 1000);
```
Comment (bobtseattle) [inline src/main/java/frc/team2412/robot/subsystem/ClimbSubsystem.java:63]: 60, 60?  Aren't the breakers and the PDH limited to a maximum of 40 Amps?

**Review by bobtseattle — COMMENTED** [review]
If should enable the PDH shuffleboard plugin so we know how much power draw is really necessary before setting values that may run the risk of brownouts and popping breakers in the middle of a match.

File: src/main/java/frc/team2412/robot/subsystem/ClimbSubsystem.java (hunk 11 above)
Comment (SumedhP) [reply src/main/java/frc/team2412/robot/subsystem/ClimbSubsystem.java:63]: The breakers are able sustain 40amps indefinitely but can go above it temporarily. I've been told 150% for 15 seconds is allowed

Comment (bobtseattle) [reply src/main/java/frc/team2412/robot/subsystem/ClimbSubsystem.java:63]: But it doesn't seem like this ever limits the motor back down to 40a, even after the limit triggers. 
Shouldn't this be along the lines of 40, 60, 15?

Comment (TAKBS2412) [reply src/main/java/frc/team2412/robot/subsystem/ClimbSubsystem.java:63]: Yeah this seems a bit off, especially considering that the time units are in seconds, and 1000 seconds is longer than an entire match.

File: src/main/java/frc/team2412/robot/subsystem/ClimbSubsystem.java [hunk 12]
```diff
@@ -145,12 +182,24 @@ public void angleClimbHook(DoubleSolenoid.Value value) {
         solenoid.set(value);
     }
 
+    @Config(name = "Stop Fixed Motor")
+    public void stopFixedPID(boolean stop) {
+        if (stop) {
+            climbFixedMotor.stopMotor();
+        }
+        ;
```
Comment (TAKBS2412) [inline src/main/java/frc/team2412/robot/subsystem/ClimbSubsystem.java:190]: remove this line
